src/fsharp/FSharp.Core/control.fs

```diff
@@ -84,6 +84,33 @@ namespace Microsoft.FSharp.Control
             member this.Dispose() = this.Dispose()
 
     
+    /// Global mutable state used to associate Exception
+
+    [<AutoOpen>]
+    module ExceptionDispatchInfoHelpers =
+
+        let associationTable = System.Runtime.CompilerServices.ConditionalWeakTable<exn, ExceptionDispatchInfo>()
+
+        type ExceptionDispatchInfo with 
+
+            member edi.GetAssociatedSourceException() = 
+                let exn = edi.SourceException
+                // Try to store the entry in the association table to allow us to recover it later.
+                try lock associationTable (fun () -> associationTable.Add(exn, edi)) with _ -> ()
```

Comment:
I believe CWTs are thread safe and do not require locking

Reply:
Thanks, you're correct